llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -4247,6 +4247,19 @@ SDValue DAGCombiner::visitSUB(SDNode *N) {
       sd_match(N1, m_UMaxLike(m_Specific(A), m_Specific(B))))
     return DAG.getNegative(DAG.getNode(ISD::ABDU, DL, VT, A, B), DL, VT);
 
+  // (sub x, (select (ult x, y), 0, y)) -> (umin x, (sub x, y))
+  auto LK = TLI.getTypeConversion(*DAG.getContext(), VT);
+  if ((LK.first == TargetLoweringBase::TypeLegal ||
+       LK.first == TargetLoweringBase::TypePromoteInteger) &&
+      TLI.isOperationLegal(ISD::UMIN, LK.second)) {
+    SDValue Y;
+    if (sd_match(N1, m_OneUse(m_Select(m_SetCC(m_Specific(N0), m_Value(Y),
+                                               m_SpecificCondCode(ISD::SETULT)),
```

Comment:
are you planning to the `(select (uge y, x), y, 0)` case, which @topperc also mentioned, in this PR too?

Reply:
I was going to -- that's the reason why I turned this PR into a draft. But InstCombine seems to avoid this pattern.

Reply:
If the setcc has another user will it prevent InstCombine from canonicalizing it?

Reply:
It won't. This is covered by the `sub_if_uge_multiuse_cmp_i32` test ([Compiler Explorer](https://godbolt.org/z/Ga1vjWx3T)). This test shows that the transform is profitable even if setcc has another user.

Reply:
@topperc I'm waiting for your approval.

Reply:
Here is a test where InstCombine doesn't change the uge to ult https://godbolt.org/z/oTafcozaG

Reply:
Added. Please review.